lisp/tree/frame.lisp

```diff
@@ -133,14 +133,12 @@
 (defun swap-in-parent (frame new-parent)
   "In FRAME's parent, swap FRAME for NEW-PARENT. Don't change the dimensions of FRAME."
   ;; TODO: also swap the parent pointer in NEW-PARENT
-  (if (root-frame-p frame)
-      (setf (root-tree (frame-parent frame)) new-parent)
-      (replace-item (tree-children (frame-parent frame)) frame new-parent)))
+  (replace-item (tree-children (frame-parent frame)) frame new-parent))
 
 (defun %replace-frame (root frame)
   "Replace ROOT with FRAME without any cleanup. Change the dimensions and position
 of FRAME to those of ROOT."
-  (if (root-frame-p root)
+  (if (and (root-frame-p root) (not (cdr (tree-children (frame-parent (frame-parent root))))))
```

Comment:
At least add a comment explaining this weird condition, but there might be a more abstract way of checking this.